java-checks/src/main/java/org/sonar/java/checks/spring/SpringBeansShouldBeAccessibleCheck.java

```diff
@@ -0,0 +1,157 @@
+
+  private static final String MESSAGE_FORMAT = "'%s' is not reachable by @ComponentsScan or @SpringBootApplication. "
+    + "Either move it to a package configured in @ComponentsScan or update your @ComponentsScan configuration.";
+
+  private static final String[] SPRING_BEAN_ANNOTATIONS = {
+    "org.springframework.stereotype.Component",
+    "org.springframework.stereotype.Service",
+    "org.springframework.stereotype.Repository",
+    "org.springframework.stereotype.Controller",
+    "org.springframework.web.bind.annotation.RestController"
+  };
+
+  private static final String COMPONENT_SCAN_ANNOTATION = "org.springframework.context.annotation.ComponentScan";
+  private static final Set<String> COMPONENT_SCAN_ARGUMENTS = new HashSet<>(Arrays.asList("basePackages", "value"));
+
+  private static final String SPRING_BOOT_APP_ANNOTATION = "org.springframework.boot.autoconfigure.SpringBootApplication";
+
+  private final Map<String, List<AnalyzerMessage>> messagesPerPackage = new HashMap<>();
+  private final Set<String> scannedPackages = new HashSet<>();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.CLASS);
+  }
+
+  @Override
+  public void endOfAnalysis() {
+    if (!(context instanceof AnalyzerMessageReporter)) {
+      return;
+    }
+
+    AnalyzerMessageReporter reporter = (AnalyzerMessageReporter) context;
+    messagesPerPackage.entrySet().stream()
+      // also consider sub-packages
+      .filter(entry -> scannedPackages.stream().noneMatch(entry.getKey()::contains))
+      .forEach(entry -> entry.getValue().forEach(reporter::reportIssue));
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic() || !(context instanceof AnalyzerMessageReporter)) {
+      return;
+    }
+
+    ClassTree classTree = (ClassTree) tree;
+    String classPackageName = packageNameOf(classTree);
+
+    Optional<AnnotationTree> componentScanAnnotation = classTree.modifiers().annotations().stream()
+      .filter(SpringBeansShouldBeAccessibleCheck::isComponentScan).findFirst();
+    if (componentScanAnnotation.isPresent()) {
+      componentScanAnnotation.get().arguments().forEach(this::addToScannedPackages);
+    } else if (hasAnnotation(classTree, SPRING_BOOT_APP_ANNOTATION)) {
+      scannedPackages.add(classPackageName);
+    } else if (hasAnnotation(classTree, SPRING_BEAN_ANNOTATIONS)) {
+      addMessageToMap(classPackageName, classTree);
+    }
+  }
+
+  private void addMessageToMap(String classPackageName, ClassTree classTree) {
+    AnalyzerMessageReporter reporter = (AnalyzerMessageReporter) context;
+    IdentifierTree className = classTree.simpleName();
+    if (className != null) {
+      AnalyzerMessage analyzerMessage = reporter.createAnalyzerMessage(this, className, String.format(MESSAGE_FORMAT, className));
+      messagesPerPackage.computeIfAbsent(classPackageName, k -> new ArrayList<>());
```

Comment:
`computeIfAbsent` returns the value if present, or the new value. So you already have the list and does not need to get it again on the next line. Just chain the `.add(analyzerMessage)`